File: modules/checks/hostinjection.py
"""Host header injection checks.

Tests for:
1. Host header reflection — injected value appears in response body or headers
   (cache poisoning, open redirect, link injection vectors)
2. Password reset poisoning — forged Host causes reset links to point to
   attacker-controlled domain; check verifies server reflects the injected host
   in response body (email content preview / token in response)
3. X-Forwarded-Host / X-Host / proxy header variants
"""
from __future__ import annotations
import re
import requests

from ..findings import Finding, Severity
from ..session import session_to_curl_flags

PROBE_HOST = "vectrix-probe.attacker-host.com"

HOST_OVERRIDE_HEADERS = [
    "X-Forwarded-Host",
    "X-Host",
    "X-Forwarded-Server",
    "X-HTTP-Host-Override",
    "X-Original-Host",
]

RESET_PATHS = [
    "/forgot-password", "/forgot_password", "/password/reset",
    "/password-reset", "/auth/reset", "/api/password/reset",
    "/api/auth/forgot", "/api/auth/password/reset",
    "/users/password", "/account/forgot", "/reset-password",
    "/api/v1/auth/forgot", "/api/v2/auth/forgot",
]


# ── Main entry ────────────────────────────────────────────────────────────────

def run(session: requests.Session, base_url: str, endpoints: list, store) -> None:
    print("[*] Checking Host header injection...")
    curl_auth = session_to_curl_flags(session)
    timeout = getattr(session, 'timeout', 15)

    _check_host_reflection(session, base_url, curl_auth, timeout, store)
    _check_reset_poisoning(session, base_url, endpoints, curl_auth, timeout, store)


# ── Host reflection ───────────────────────────────────────────────────────────

def _check_host_reflection(session, base_url, curl_auth, timeout, store):
    """Try injecting PROBE_HOST via various override headers; look for reflection."""
    for header in HOST_OVERRIDE_HEADERS:
        try:
            resp = session.get(
                base_url,
                headers={header: PROBE_HOST},
                timeout=timeout,
                allow_redirects=False,
            )
        except Exception:
            continue

        probe_lower = PROBE_HOST.lower()
        reflected_in_body = probe_lower in resp.text.lower()
        reflected_in_location = probe_lower in resp.headers.get("Location", "").lower()
        reflected_in_link = probe_lower in resp.headers.get("Link", "").lower()

        if not (reflected_in_body or reflected_in_location or reflected_in_link):
            continue

        where = []
        if reflected_in_body:      where.append("body")
        if reflected_in_location:  where.append("Location header")
        if reflected_in_link:      where.append("Link header")

        store.add(Finding(
            title=f"Host Header Injection — отражение через '{header}'",
            severity=Severity.HIGH,
            category="Injection",
            cwe="CWE-113",
            description=(
                f"Заголовок '{header}: {PROBE_HOST}' отражается в ответе ({', '.join(where)}).\n\n"
                "Векторы эксплуатации:\n"
                "• **Web Cache Poisoning** — CDN/proxy кэшируют ответ с injected-ссылками\n"
                "• **Password Reset Poisoning** — ссылка в письме указывает на сервер атакующего\n"
                "• **Open Redirect** — Location содержит внешний домен\n"
                "• **SSRF через Host** — внутренние сервисы принимают запросы по поддельному хосту"
            ),
            url=base_url,
            evidence=(
                f"Заголовок: {header}: {PROBE_HOST}\n"
                f"Зеркало найдено в: {', '.join(where)}\n"
                f"HTTP {resp.status_code}"
            ),
            remediation=(
                "1. Формируйте абсолютные URL из жёстко заданного домена в конфигурации, "
                "   а не из заголовка HTTP Host.\n"
                "2. Валидируйте Host и X-Forwarded-Host по whitelist разрешённых доменов.\n"
                "3. В nginx/Apache/CDN задайте canonical hostname и отклоняйте нераспознанные.\n"
                "4. Для cache poisoning: установьте Vary: Host на кэшируемые ответы."
            ),
            reproduction=(
                f"curl -sk {curl_auth} -H '{header}: {PROBE_HOST}' -I '{base_url}'"
            ),
        ))
        return  # One finding per base_url is enough


# ── Password reset poisoning ──────────────────────────────────────────────────

def _check_reset_poisoning(session, base_url, endpoints, curl_auth, timeout, store):
    """Try to trigger a reset and check if the injected Host appears in response."""
    # Collect reset endpoints from crawl + static list
    reset_urls: set[str] = set()
    for ep in endpoints:
        if any(p in ep.url.lower() for p in RESET_PATHS):
            reset_urls.add(ep.url)
    for path in RESET_PATHS:
        reset_urls.add(base_url.rstrip("/") + path)

    for url in list(reset_urls)[:12]:
        # Quick GET probe — skip if 404/410
        try:
            probe = session.get(url, timeout=timeout, allow_redirects=False)
            if probe.status_code in (404, 410):
                continue
        except Exception:
            continue

        for header in ["Host", "X-Forwarded-Host"]:
            try:
                extra = {header: PROBE_HOST}
                if header == "Host":
                    extra["X-Forwarded-Host"] = PROBE_HOST

                resp = session.post(
                    url,
                    data={"email": "pentest-probe@example.com"},
                    headers=extra,
                    timeout=timeout,
                    allow_redirects=False,
                )
            except Exception:
                continue

            if resp.status_code not in (200, 201, 202, 204, 302, 301):
                continue

            if PROBE_HOST.lower() not in resp.text.lower():
                continue

            store.add(Finding(
                title="Password Reset Poisoning — Host header в теле ответа сброса пароля",
                severity=Severity.HIGH,
                category="Authentication",
                cwe="CWE-640",
                description=(
                    f"Endpoint {url} принял POST с поддельным {header} "
                    f"и отразил '{PROBE_HOST}' в теле ответа.\n\n"
                    "Если сервер использует Host для формирования ссылки в reset-письме, "
                    "атакующий может перехватить токен сброса пароля:\n"
                    "1. Злоумышленник инициирует сброс для жертвы с Host: evil.com\n"
                    "2. Жертва получает письмо со ссылкой https://evil.com/reset?token=...\n"
                    "3. При клике — токен улетает на сервер атакующего\n"
                    "4. Атакующий меняет пароль жертвы"
                ),
                url=url,
                method="POST",
                evidence=(
                    f"{header}: {PROBE_HOST} → '{PROBE_HOST}' найдено в ответе\n"
                    f"HTTP {resp.status_code}"
                ),
                remediation=(
                    "1. Генерируйте reset-ссылки с доменом из конфигурации приложения, "
                    "   НЕ из заголовка Host.\n"
                    "2. Добавьте whitelist разрешённых доменов для Host/X-Forwarded-Host.\n"
                    "3. Проверьте: config('app.url') в Laravel, settings.ALLOWED_HOSTS в Django, "
                    "   ServerName в Apache, server_name в nginx."
                ),
                reproduction=(
                    f"curl -sk {curl_auth} -X POST "
                    f"-H 'Host: {PROBE_HOST}' "
                    f"-H 'X-Forwarded-Host: {PROBE_HOST}' "
                    f"-d 'email=victim@target.com' '{url}'"
                ),
            ))
            return  # One finding is enough

File: go-vectrix/internal/checks/hostinjection.go
package checks

import (
	"fmt"
	"net/url"
	"strings"
	"time"

	"vectrixgo/internal/crawler"
	"vectrixgo/internal/findings"
	"vectrixgo/internal/httpsession"
	"vectrixgo/internal/logging"
)

// hostinjProbeHost mirrors PROBE_HOST in hostinjection.py.
const hostinjProbeHost = "vectrix-probe.attacker-host.com"

// hostinjOverrideHeaders mirrors HOST_OVERRIDE_HEADERS.
var hostinjOverrideHeaders = []string{
	"X-Forwarded-Host",
	"X-Host",
	"X-Forwarded-Server",
	"X-HTTP-Host-Override",
	"X-Original-Host",
}

// hostinjResetPaths mirrors RESET_PATHS.
var hostinjResetPaths = []string{
	"/forgot-password", "/forgot_password", "/password/reset",
	"/password-reset", "/auth/reset", "/api/password/reset",
	"/api/auth/forgot", "/api/auth/password/reset",
	"/users/password", "/account/forgot", "/reset-password",
	"/api/v1/auth/forgot", "/api/v2/auth/forgot",
}

// RunHostInjection checks for Host header reflection (cache poisoning, open
// redirect) and password-reset poisoning via forged Host headers. Mirrors
// modules/checks/hostinjection.py run().
func RunHostInjection(session *httpsession.Session, baseURL string, endpoints []crawler.Endpoint, store *findings.FindingStore) {
	logging.Println("[*] Checking host header injection...")
	curlAuth := session.CurlAuthFlags(baseURL)

	hostinjCheckHostReflection(session, baseURL, curlAuth, store)
	hostinjCheckResetPoisoning(session, baseURL, endpoints, curlAuth, store)
}

// hostinjCheckHostReflection mirrors _check_host_reflection — tries injecting
// PROBE_HOST via various override headers and looks for reflection.
func hostinjCheckHostReflection(session *httpsession.Session, baseURL, curlAuth string, store *findings.FindingStore) {
	for _, header := range hostinjOverrideHeaders {
		resp, err := session.Request("GET", baseURL, httpsession.Options{
			Headers:        map[string]string{header: hostinjProbeHost},
			AllowRedirects: false,
			Timeout:        15 * time.Second,
		})
		if err != nil {
			continue
		}

		probeLower := strings.ToLower(hostinjProbeHost)
		reflectedInBody := strings.Contains(strings.ToLower(resp.Body), probeLower)
		reflectedInLocation := strings.Contains(strings.ToLower(resp.Header.Get("Location")), probeLower)
		reflectedInLink := strings.Contains(strings.ToLower(resp.Header.Get("Link")), probeLower)

		if !reflectedInBody && !reflectedInLocation && !reflectedInLink {
			continue
		}

		var where []string
		if reflectedInBody {
			where = append(where, "body")
		}
		if reflectedInLocation {
			where = append(where, "Location header")
		}
		if reflectedInLink {
			where = append(where, "Link header")
		}

		f := findings.NewFinding(
			fmt.Sprintf("Host Header Injection — отражение через '%s'", header),
			findings.High,
			"Injection",
			"CWE-113",
			baseURL,
			fmt.Sprintf("Заголовок '%s: %s' отражается в ответе (%s).\n\n"+
				"Векторы эксплуатации:\n"+
				"• **Web Cache Poisoning** — CDN/proxy кэшируют ответ с injected-ссылками\n"+
				"• **Password Reset Poisoning** — ссылка в письме указывает на сервер атакующего\n"+
				"• **Open Redirect** — Location содержит внешний домен\n"+
				"• **SSRF через Host** — внутренние сервисы принимают запросы по поддельному хосту",
				header, hostinjProbeHost, strings.Join(where, ", ")),
			"1. Формируйте абсолютные URL из жёстко заданного домена в конфигурации, "+
				"   а не из заголовка HTTP Host.\n"+
				"2. Валидируйте Host и X-Forwarded-Host по whitelist разрешённых доменов.\n"+
				"3. В nginx/Apache/CDN задайте canonical hostname и отклоняйте нераспознанные.\n"+
				"4. Для cache poisoning: установьте Vary: Host на кэшируемые ответы.",
			fmt.Sprintf("curl -sk %s -H '%s: %s' -I '%s'", curlAuth, header, hostinjProbeHost, baseURL),
		)
		f.Evidence = fmt.Sprintf("Заголовок: %s: %s\nЗеркало найдено в: %s\nHTTP %d",
			header, hostinjProbeHost, strings.Join(where, ", "), resp.StatusCode)
		store.Add(f)
		return // One finding per base_url is enough
	}
}

// hostinjCheckResetPoisoning mirrors _check_reset_poisoning — tries to
// trigger a password reset and checks if the injected Host appears in the
// response.
func hostinjCheckResetPoisoning(session *httpsession.Session, baseURL string, endpoints []crawler.Endpoint, curlAuth string, store *findings.FindingStore) {
	// Collect reset endpoints from crawl + static list.
	resetURLs := make(map[string]struct{})
	for _, ep := range endpoints {
		urlLower := strings.ToLower(ep.URL)
		for _, p := range hostinjResetPaths {
			if strings.Contains(urlLower, p) {
				resetURLs[ep.URL] = struct{}{}
				break
			}
		}
	}
	for _, path := range hostinjResetPaths {
		resetURLs[strings.TrimRight(baseURL, "/")+path] = struct{}{}
	}

	count := 0
	for u := range resetURLs {
		if count >= 12 {
			break
		}
		count++

		// Quick GET probe — skip if 404/410.
		probe, err := session.Request("GET", u, httpsession.Options{AllowRedirects: false, Timeout: 15 * time.Second})
		if err != nil {
			continue
		}
		if probe.StatusCode == 404 || probe.StatusCode == 410 {
			continue
		}

		for _, header := range []string{"Host", "X-Forwarded-Host"} {
			extra := map[string]string{header: hostinjProbeHost}
			if header == "Host" {
				extra["X-Forwarded-Host"] = hostinjProbeHost
			}

			form := url.Values{"email": {"pentest-probe@example.com"}}
			h := map[string]string{"Content-Type": "application/x-www-form-urlencoded"}
			for k, v := range extra {
				h[k] = v
			}

			resp, err := session.Request("POST", u, httpsession.Options{
				Headers:        h,
				Body:           strings.NewReader(form.Encode()),
				AllowRedirects: false,
				Timeout:        15 * time.Second,
			})
			if err != nil {
				continue
			}

			switch resp.StatusCode {
			case 200, 201, 202, 204, 302, 301:
			default:
				continue
			}

			if !strings.Contains(strings.ToLower(resp.Body), strings.ToLower(hostinjProbeHost)) {
				continue
			}

			f := findings.NewFinding(
				"Password Reset Poisoning — Host header в теле ответа сброса пароля",
				findings.High,
				"Authentication",
				"CWE-640",
				u,
				fmt.Sprintf("Endpoint %s принял POST с поддельным %s "+
					"и отразил '%s' в теле ответа.\n\n"+
					"Если сервер использует Host для формирования ссылки в reset-письме, "+
					"атакующий может перехватить токен сброса пароля:\n"+
					"1. Злоумышленник инициирует сброс для жертвы с Host: evil.com\n"+
					"2. Жертва получает письмо со ссылкой https://evil.com/reset?token=...\n"+
					"3. При клике — токен улетает на сервер атакующего\n"+
					"4. Атакующий меняет пароль жертвы",
					u, header, hostinjProbeHost),
				"1. Генерируйте reset-ссылки с доменом из конфигурации приложения, "+
					"   НЕ из заголовка Host.\n"+
					"2. Добавьте whitelist разрешённых доменов для Host/X-Forwarded-Host.\n"+
					"3. Проверьте: config('app.url') в Laravel, settings.ALLOWED_HOSTS в Django, "+
					"   ServerName в Apache, server_name в nginx.",
				fmt.Sprintf("curl -sk %s -X POST -H 'Host: %s' -H 'X-Forwarded-Host: %s' -d 'email=victim@target.com' '%s'",
					curlAuth, hostinjProbeHost, hostinjProbeHost, u),
			)
			f.Method = "POST"
			f.Evidence = fmt.Sprintf("%s: %s → '%s' найдено в ответе\nHTTP %d",
				header, hostinjProbeHost, hostinjProbeHost, resp.StatusCode)
			store.Add(f)
			return // One finding is enough
		}
	}
}
